ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java

```diff
@@ -26,36 +26,104 @@
 import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
 import io.confluent.kafka.serializers.AbstractKafkaSchemaSerDeConfig;
 import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.model.WindowType;
 import io.confluent.ksql.parser.DurationParser;
 import io.confluent.ksql.query.QueryId;
 import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.query.QuerySchemas.SchemaInfo;
 import io.confluent.ksql.serde.FormatFactory;
 import io.confluent.ksql.serde.KeyFormat;
 import io.confluent.ksql.serde.ValueFormat;
+import io.confluent.ksql.serde.WindowInfo;
 import io.confluent.ksql.test.TestFrameworkException;
 import io.confluent.ksql.test.serde.SerdeSupplier;
 import io.confluent.ksql.test.utils.SerdeUtil;
 import io.confluent.ksql.util.PersistentQueryMetadata;
+import java.time.Duration;
 import java.util.List;
+import java.util.Optional;
 import java.util.OptionalLong;
 import java.util.Set;
+import java.util.function.Function;
+import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import org.apache.kafka.common.serialization.Deserializer;
 import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.streams.TopologyDescription;
+import org.apache.kafka.streams.TopologyDescription.Processor;
 import org.apache.kafka.streams.kstream.TimeWindowedDeserializer;
 
 /**
  * Cache of info known about topics in use in the test.
+ *
+ * <p>Info for source and sink topics is obtained by finding a {@link
+ * io.confluent.ksql.metastore.model.DataSource} with a matching source topic name in the {@link
+ * io.confluent.ksql.metastore.MetaStore}.
+ *
+ * <p>Info for internal topics is obtained from the {@link PersistentQueryMetadata#getSchemas()}.
+ * This is a map of {@code loggerNamePrefix} to {@link SchemaInfo}. This map is populated as a query
+ * is built, so presents the <i>actual</i> schema and formats used. This class uses pattern matching
+ * against the topic name to determine the correct {@code loggerNamePrefix} to look up and any
+ * additional logic needded.
  */
 public class TopicInfoCache {
 
-  private static final Pattern INTERNAL_TOPIC_PATTERN = Pattern
-      .compile("_confluent.*query_(.*_\\d+)-.*-(changelog|repartition)");
+  private static final String TOPIC_PATTERN_PREFIX = "_confluent.*query_(?<queryId>.*_\\d+)-";
+
+  private static final List<InternalTopicPattern> INTERNAL_TOPIC_PATTERNS = ImmutableList.of(
```

Comment:
this class seems super fragile... we're extracting information from internal topic naming conventions, mapping that to logger names, which we use to extract the serde info that we pass to Kafka Streams, and then magically deduce from that what the actual serde should be based on what we know Kafka Streams does. I'm sure that coding this up you had a bad feeling in your stomach, this just doesn't sit right! Can we do something else?

What we need is a mapping from topic name to logger prefix. We could have the `LoggingSerializer` aware of the prefix that it's using, and register this mapping whenever it (de)serializes anything (and it's easy to pass in the prefix there, we have it in `GenericSerdeFactory`). Since we probably don't want to maintain this mapping in PROD as it might have real perf implications to do this per-message, we could do the evil thing of having a static map that is populated and cleared just for tests.

I don't love that solution either, but maybe it'll get the brain juices flowing and I think it's a lot more robust (even if it breaks some encapsulation) than what's here.

Reply:
You're right, of course. Was just trying to get this done so could move on, but looking,...

Reply:
Your wish is my command...